File: DSAL/A-CPP/A3.cpp
// 3. For given set of elements create skip list. Find the element in the set
// that is closest to some given value. (Note: Decide the level of element in
// the list Randomly with some upper limit)

#include <iostream>
#include <string>
using namespace std;

class SimpleRandom {
 private:
  unsigned long long state;

 public:
  explicit SimpleRandom(unsigned long long seed) {
    if (seed == 0) seed = 1;
    state = seed;
  }

  unsigned int next() {
    state = (state * 48271ULL) % 2147483647ULL;
    return static_cast<unsigned int>(state);
  }

  int nextBit() { return static_cast<int>(next() % 2U); }
};

class SkipList {
 private:
  struct Node {
    int key;
    int level;
    Node **forward;

    Node(int k, int lvl) {
      key = k;
      level = lvl;
      forward = new Node *[lvl + 1];
      for (int i = 0; i <= lvl; ++i) {
        forward[i] = 0;
      }
    }

    ~Node() { delete[] forward; }
  };

  int maxLevel;
  int currentLevel;
  Node *header;
  SimpleRandom rng;

  int absInt(int x) const { return (x < 0) ? -x : x; }

  int randomLevel() {
    int lvl = 0;
    while (lvl < maxLevel && rng.nextBit() == 1) {
      ++lvl;
    }
    return lvl;
  }

 public:
  SkipList(int maxLvl, unsigned long long seed)
      : maxLevel(maxLvl), currentLevel(0), header(new Node(0, maxLvl)), rng(seed) {}

  ~SkipList() {
    Node *cur = header;
    while (cur != 0) {
      Node *nxt = cur->forward[0];
      delete cur;
      cur = nxt;
    }
  }

  bool insert(int key) {
    Node **update = new Node *[maxLevel + 1];
    Node *cur = header;

    for (int i = currentLevel; i >= 0; --i) {
      while (cur->forward[i] != 0 && cur->forward[i]->key < key) {
        cur = cur->forward[i];
      }
      update[i] = cur;
    }

    cur = cur->forward[0];
    if (cur != 0 && cur->key == key) {
      delete[] update;
      return false;
    }

    int nodeLevel = randomLevel();
    if (nodeLevel > currentLevel) {
      for (int i = currentLevel + 1; i <= nodeLevel; ++i) {
        update[i] = header;
      }
      currentLevel = nodeLevel;
    }

    Node *node = new Node(key, nodeLevel);
    for (int i = 0; i <= nodeLevel; ++i) {
      node->forward[i] = update[i]->forward[i];
      update[i]->forward[i] = node;
    }

    delete[] update;
    return true;
  }

  bool findClosest(int key, int &closest) const {
    Node *cur = header;

    for (int i = currentLevel; i >= 0; --i) {
      while (cur->forward[i] != 0 && cur->forward[i]->key < key) {
        cur = cur->forward[i];
      }
    }

    Node *left = (cur == header) ? 0 : cur;
    Node *right = cur->forward[0];

    if (left == 0 && right == 0) return false;
    if (left == 0) {
      closest = right->key;
      return true;
    }
    if (right == 0) {
      closest = left->key;
      return true;
    }

    int dl = absInt(left->key - key);
    int dr = absInt(right->key - key);
    closest = (dl <= dr) ? left->key : right->key;
    return true;
  }

  void display() const {
    cout << "\nSkip List Structure\n";
    for (int lvl = currentLevel; lvl >= 0; --lvl) {
      cout << "Level " << lvl << " : ";
      Node *cur = header->forward[lvl];
      while (cur != 0) {
        cout << cur->key << " ";
        cur = cur->forward[lvl];
      }
      cout << "\n";
    }
  }
};

int main() {
  int maxLvl;
  unsigned long long seed;

  cout << "Enter max level for skip list: ";
  cin >> maxLvl;
  if (maxLvl < 1) {
    cout << "Max level must be at least 1\n";
    return 0;
  }

  cout << "Enter random seed (any positive integer): ";
  cin >> seed;
  if (seed == 0) seed = 1;

  SkipList sl(maxLvl, seed);

  while (true) {
    cout << "\n1. Insert Element\n";
    cout << "2. Find Closest Element\n";
    cout << "3. Display List\n";
    cout << "4. Exit\n";
    cout << "Enter choice: ";

    int choice;
    cin >> choice;

    if (choice == 1) {
      int value;
      cout << "Enter integer element: ";
      cin >> value;
      if (sl.insert(value)) cout << "Inserted " << value << "\n";
      else cout << "Element already exists\n";

    } else if (choice == 2) {
      int value, closest;
      cout << "Enter value to find closest: ";
      cin >> value;
      if (sl.findClosest(value, closest)) {
        cout << "Closest element to " << value << " is " << closest << "\n";
      } else {
        cout << "Skip list is empty\n";
      }

    } else if (choice == 3) {
      sl.display();

    } else if (choice == 4) {
      cout << "Exiting...\n";
      break;

    } else {
      cout << "Invalid choice\n";
    }
  }

  return 0;
}

File: DSAL/A/A3.py
# 3. For given set of elements create skip list. Find the element in the set that is closest to some given value. (Note: Decide the level of element in the list Randomly with some upper limit)

class Random:
    def __init__(self, seed=None):
        if seed is None:
            self.state = id(object()) % 2147483647
        else:
            self.state = seed
        if self.state == 0:
            self.state = 1

    def rand(self):
        a = 48271
        m = 2147483647
        self.state = (a * self.state) % m
        return self.state

    def random_float(self):
        return self.rand() / 2147483647.0

class Node:
    def __init__(self, key, level):
        self.key = key
        self.forward = [None] * (level + 1)

class SkipList:
    def __init__(self, max_level, p):
        self.max_level = max_level
        self.p = p
        self.header = Node(-1, max_level)
        self.level = 0
        self.random = Random()

    def random_level(self):
        level = 0
        while self.random.random_float() < self.p and level < self.max_level:
            level += 1
        return level

    def insert(self, key):
        update = [None] * (self.max_level + 1)
        current = self.header

        for i in range(self.level, -1, -1):
            while current.forward[i] is not None and current.forward[i].key < key:
                current = current.forward[i]
            update[i] = current

        current = current.forward[0]

        if current is None or current.key != key:
            r_level = self.random_level()

            if r_level > self.level:
                for i in range(self.level + 1, r_level + 1):
                    update[i] = self.header
                self.level = r_level

            new_node = Node(key, r_level)

            for i in range(r_level + 1):
                new_node.forward[i] = update[i].forward[i]
                update[i].forward[i] = new_node
            
            print("Inserted", key, "at level", r_level)
        else:
            print("Key", key, "already exists")

    def find_closest(self, key):
        current = self.header
        
        for i in range(self.level, -1, -1):
            while current.forward[i] is not None and current.forward[i].key < key:
                current = current.forward[i]
        
        candidate1 = current
        candidate2 = current.forward[0]

        dist1 = float('inf')
        dist2 = float('inf')

        if candidate1.key != -1:
            dist1 = abs(candidate1.key - key)
        
        if candidate2 is not None:
            dist2 = abs(candidate2.key - key)

        if dist1 == float('inf') and dist2 == float('inf'):
            return None
        
        if dist1 <= dist2:
            return candidate1.key
        else:
            return candidate2.key

    def display(self):
        print("\nSkip List Structure:")
        for i in range(self.level, -1, -1):
            head = self.header.forward[i]
            print("Level", i, ":", end=" ")
            while head is not None:
                print(head.key, end=" ")
                head = head.forward[i]
            print()

def main():
    max_lvl = int(input("Enter max level for skip list: "))
    probability = 0.5
    
    sl = SkipList(max_lvl, probability)

    while True:
        print("\n1. Insert Element")
        print("2. Find Closest Element")
        print("3. Display List")
        print("4. Exit")
        
        choice = input("Enter choice: ")

        if choice == '1':
            val = int(input("Enter integer element: "))
            sl.insert(val)

        elif choice == '2':
            val = int(input("Enter value to find closest: "))
            result = sl.find_closest(val)
            if result is not None:
                print("Closest element to", val, "is", result)
            else:
                print("List is empty")

        elif choice == '3':
            sl.display()

        elif choice == '4':
            break

        else:
            print("Invalid choice")

if __name__ == "__main__":
    main()